File: singly_linked_list/recipe-01/c/main.c
#include "singly_linked_list_string.h"
#include <stdio.h>

void print1(ItemType data) {
    printf("Got this data: %s\n", data);
}

void print2(ItemType data) {
    printf("data: %s\n", data);
}

void test()
{
    printf("----------------------------%s---------------------------\n", __func__);

    SinglyLinkedList* words = NewList();
    Append(words, "foo");
    Append(words, "bar");
    Append(words, "bim");
    Append(words, "baz");
    Append(words, "quux");

    printf("access by index\n");
    printf("here is a node: %s\n", GetItem(words, 1));
    
    printf("modify by index\n");
    SetItem(words, 4, "Quux", &SetData);
    printf("Modified node by index: %s\n", GetItem(words, 4));

    printf("This list has %d elements.\n", Count(words));
    Travel(words, &print1);

    if (Search(words, "foo", &IsEqual))
        printf("Found foo in the list.\n");
    if (Search(words, "amiga", &IsEqual))
        printf("Found amiga in the list.\n");
    
    printf("Now we try to Delete an item\n");
    Delete(words, "bim", &IsEqual);
    printf("List now has %d elements\n", Count(words));
    Travel(words, &print2);
    
    printf("Delete the first item in the list\n");
    Delete(words, "foo", &IsEqual);
    printf("size: %d\n", Count(words));
    Travel(words, &print2);
    
    printf("Delete the last item in the list\n");
    Delete(words, "quux", &IsEqual);
    printf("size: %d\n", Count(words));
    Travel(words, &print2);

    FreeList(words);
}

void test_reverse()
{
    printf("----------------------------%s---------------------------\n", __func__);

    SinglyLinkedList* words = NewList();
    Append(words, "foo");
    Append(words, "bar");
    Append(words, "bim");
    Append(words, "baz");
    Append(words, "quux");

    printf("The origin list\n");
    Travel(words, &print2);

    printf("Reverse the list\n");
    Reverse(words);
    Travel(words, &print2);

    FreeList(words);
}

int main()
{
    test();
    test_reverse();

    return 0;
}

File: singly_linked_list/recipe-01/python/main.py
from singly_linked_list import SinglyLinkedList

def test():
    print("----------------------------test---------------------------");

    words = SinglyLinkedList()
    words.Append('foo')
    words.Append('bar')
    words.Append('bim')
    words.Append('baz')
    words.Append('quux')
    
    print("access by index")
    print("here is a node: {}".format(words[1]))
    
    print("modify by index")
    words[4] = "Quux"
    print("Modified node by index: {}".format(words[4]))
    
    print("This list has {} elements.".format(words.count))
    for word in words:
        print("Got this data: {}".format(word))
    
    if words.Search('foo'):
        print("Found foo in the list.")
    if words.Search('amiga'):
        print("Found amiga in the list.")
    
    print("Now we try to Delete an item")
    words.Delete('bim')
    print("List now has {} elements".format(words.count))
    for word in words:
        print("data: {}".format(word))
    
    print("Delete the first item in the list")
    words.Delete('foo')
    print("size: {}".format(words.count))
    for word in words:
        print("data: {}".format(word))
    
    print("Delete the last item in the list")
    words.Delete('quux')
    print("size: {}".format(words.count))
    for word in words:
        print("data: {}".format(word))

def test_reverse():
    print("----------------------------test_reverse---------------------------");

    words = SinglyLinkedList()
    words.Append('foo')
    words.Append('bar')
    words.Append('bim')
    words.Append('baz')
    words.Append('quux')
    
    print("The origin list");
    for word in words:
        print("data: {}".format(word))

    print("Reverse the list");
    words.Reverse();
    for word in words:
        print("data: {}".format(word))

def test_sort():
    print("----------------------------test_sort---------------------------");

    words = SinglyLinkedList()
    words.Append('foo')
    words.Append('bar')
    words.Append('bim')
    words.Append('baz')
    words.Append('quux')
    
    print("The origin list");
    for word in words:
        print("data: {}".format(word))

    print("Sort the list");
    words.Sort();
    for word in words:
        print("data: {}".format(word))

def main():
    test()
    test_reverse()
    test_sort()

main()
